Storage/SQL/SQLiteHistory.swift

```diff
@@ -106,6 +106,21 @@ open class SQLiteHistory {
 private let topSitesQuery = "SELECT \(TableCachedTopSites).*, \(TablePageMetadata).provider_name FROM \(TableCachedTopSites) LEFT OUTER JOIN \(TablePageMetadata) ON \(TableCachedTopSites).url = \(TablePageMetadata).site_url ORDER BY frecencies DESC LIMIT (?)"
 
 extension SQLiteHistory: BrowserHistory {
+    public func createTempTablesForFrecencyQuery() {
+        let create_awesomebarBookmarksTempTable =
+            "CREATE TEMP TABLE \(AwesomebarBookmarksTempTable) AS " +
```

Comment:
Two suggestions:

`CREATE TEMP TABLE … IF NOT EXISTS AS SELECT … FROM \(ViewAwesomebarBookmarks)`?

Reply:
`FROM \(ViewAwesomebarBookmarks)` -> Yes to that one.

`CREATE IF NOT EXISTS` -> To create a new dataset I drop the table (line 119), then create a new table (line 120), can you clarify this suggestion?

Reply:
If you, or another developer in the next couple of years, gets anything wrong, `CREATE TEMP TABLE` will return an error if the table already exists.

I haven't traced the logic, but I expect that will cause something user-facing to break. Using `CREATE TEMP TABLE IF NOT EXISTS` means that prior existence isn't a problem, and because the table definition will only change on upgrade, and temp tables don't persist across restarts, this is totally safe.

Additionally, consider using `DELETE FROM \(AwesomebarBookmarksTempTable)` instead of actually dropping the table. `DROP TABLE` literally does exactly that _then_ removes the table from the database schema, so you can save a step by not bothering.

(And that's one step closer to maintaining this materialized view rather than rebuilding it each time.)

Reply:
> If you, or another developer in the next couple of years, gets anything wrong, CREATE TEMP TABLE will return an error if the table already exists.

That is correct. The requirement of this part of the code was that the table must be dropped and then created with fresh data, and it would be easy to track down the error in someone went and changed this logic.

(Maintainability/understandability is a risk area of this patch; agreed clarity needs to be better here. Just walk me through what would make the most sense.)

> I haven't traced the logic, but I expect that will cause something user-facing to break.

In debug it will report a bunch of errors to the log and then not display autocompletions (but that behaviour is intended in terms of the error state). Hopefully the dev notices their change broke the code (assuming they don't have a test case for their change).

> Using CREATE TEMP TABLE IF NOT EXISTS means that prior existence isn't a problem, and because the table definition will only change on upgrade, and temp tables don't persist across restarts, this is totally safe.

I am confused about this prior existence concern, that would be an exceptional error case in code –the DB is broken.
The code essentially does: 'DROP TABLE FOO; CREATE TEMP TABLE FOO', –is my code not doing that? And if it is correctly a consecutive drop+create, is your suggestion still needed? 

> Additionally, consider using DELETE FROM \(AwesomebarBookmarksTempTable) instead of actually dropping the table. DROP TABLE literally does exactly that then removes the table from the database schema, so you can save a step by not bothering.

Confused, isn't DROP is far more performant than DELETE?

Reply:
From [the docs](https://sqlite.org/lang_droptable.html)

> If foreign key constraints are enabled, a DROP TABLE command performs an implicit DELETE FROM command before removing the table from the database schema.

Reply:
> The requirement of this part of the code was that the table must be dropped and then created with fresh data…

IMO the underlying requirement is: by the time we run the first query, the table exists and contains up-to-date data.

There are several ways of doing that, ranging in efficiency from best to worst:

- Keep the table accurate as we go. No work is done now.
- Incrementally update the data somehow, here and now.
- Delete all of the rows from the table and insert new ones.
- Drop the table, recreate it, and repopulate it.
- Close and reopen the database…
- Delete and recreate the database…

The reason I suggest not dropping the table and using `IF NOT EXISTS` is twofold:

1. It's more efficient.
2. If other code changes to rely on the existence of your materialized view, and the lifecycle gets more complex than you anticipated, then drop + recreate might allow the introduction of user-facing bugs.

That is: your code is correct, but future code might not be.

Given that it's possible to be just as, or more, efficient, and also cut off the possibility of some sets of bugs, I suggest doing that.

Reply:
> IMO the underlying requirement is: by the time we run the first query, the table exists and contains up-to-date data.

Agreed.

> There are several ways of doing that, ranging in efficiency from best to worst:
>    Keep the table accurate as we go. No work is done now. / Incrementally update the data somehow, here and now.
 
ASIDE: My preferred approaches of course but the scope of changes required makes it impractical :(.

> The reason I suggest not dropping the table and using IF NOT EXISTS is twofold:
>    It's more efficient.

But DELETE is O(n) and DROP is O(1) , there are no FK constraints on the temp table.
DROP is always maximally efficient, as if no FK is O(1), otherwise O(n). 
I feel like I have a learning op here, because I reject patches for using DELETE over DROP :).

> If other code changes to rely on the existence of your materialized view, and the lifecycle gets more complex than you anticipated, then drop + recreate might allow the introduction of user-facing bugs.

I don't mind making your suggested change, as the behaviour of the non-db-error state of the code is unchanged. 
Bear with me if you don't mind while I grasp the concerns, because it seems that the  suggested changes don't change the frecency query error behaviour either. 

For the problem of the the frecency query is called without the temp table, we want to:
- reduce the probability of this happening
- reduce the impact when it does happen

This PR minimizes probability by assert() in non-release builds. 
And in release the query will fail –bookmark and history completions will not show. This is my preferred level of user impact. If the probability is acceptably low (the code has reasonable defence), then I wouldn't want to mask the bug any further (not that I am sure how we would mask it further, I suppose showing old data, but we aren't discussing doing that).

The suggested changes regarding DELETE and 'CREATE IF NOT EXISTS':
- In the case of a DB error where the temp table failed to 'DROP IF EXISTS and then CREATE': I don't think you are arguing that the change would handle a bad DB state better or reduce the chance of a DB error, which I mention to clarify your argument. 
- In the case of a code or lifecycle change introducing an error (which can only be createTempTablesForFrecencyQuery() not called prior), the suggested changes don't guard this -the creational syntax we are discussing for createTempTablesForFrecencyQuery() has no effect.

Reply:
> When the WHERE is omitted from a DELETE statement and the table being deleted has no triggers, SQLite uses an optimization to erase the entire table content without having to visit each row of the table individually. This "truncate" optimization makes the delete run much faster.

Reply:
Thanks for that info, I'll change the DROP to the DELETE as suggested.